### ISSUE TITLE
Validation is not run when using "fit" method if the validation set changes its size

### ISSUE THREAD

**Issue by AlmogDavid** [keras-team/keras#15392]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04 & Windows 10 (Reproducible on any OS)
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.5 & 2.6
- Python version: 3.8
- GPU model and memory: Nvidia RTX3070

**Describe the problem**.

When using fit method of tf.keras.Model and the validation set changes its size after each epoch (I know it's not a common practice), the validation set is not run from the 3rd epoch.

After debugging we understood that the root cause for this behavior is the flag "_use_cached_eval_dataset" which passed as True to the evaluate method inside the fit method, when setting this flag to False the code behaves as expected.

**Describe the current behavior**.

The validation set is not running after 2 epochs

**Describe the expected behavior**.

The validation set should run after each epoch

- Do you want to contribute a PR? (yes/no): No (I'm not sure what is the correct fix from Keras team perspective)

**Standalone code to reproduce the issue**.

```
import random

import tensorflow as tf
import tensorflow_datasets as tfds

(ds_train, ds_test), ds_info = tfds.load(
    'mnist',
    split=['train', 'test'],
    shuffle_files=True,
    as_supervised=True,
    with_info=True,
)

def normalize_img(image, label):
  """Normalizes images: `uint8` -> `float32`."""
  return tf.cast(image, tf.float32) / 255., label

ds_train = ds_train.map(
    normalize_img, num_parallel_calls=tf.data.experimental.AUTOTUNE)
ds_train = ds_train.cache()
ds_train = ds_train.shuffle(ds_info.splits['train'].num_examples)
ds_train = ds_train.batch(128)
ds_train = ds_train.prefetch(tf.data.experimental.AUTOTUNE)
ds_train = ds_train.repeat()

ds_test = ds_test.map(
    normalize_img, num_parallel_calls=tf.data.experimental.AUTOTUNE)
ds_test = ds_test.batch(128)
ds_test = ds_test.cache()
ds_test = ds_test.prefetch(tf.data.experimental.AUTOTUNE)

# This code block creates the randomization on the dataset size
def random_sample():
    return random.random() < 0.5
ds_test = ds_test.filter(lambda x, y: tf.py_function(random_sample, inp=[], Tout=tf.bool))

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10)
])
model.compile(
    optimizer=tf.keras.optimizers.Adam(0.001),
    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    metrics=[tf.keras.metrics.SparseCategoricalAccuracy()],
)

# Uncomment to verify the validation set changes its size after each epoch
# while True:
#     for i, _ in enumerate(ds_test.as_numpy_iterator()):
#         pass
#     print(i)

model.fit(
    ds_train,
    steps_per_epoch=100,
    epochs=10,
    validation_data=ds_test,
    verbose=1
)
```

**Comment by kumariko** [comment]
@jvishnuvardhan  I could reproduce the issue with TF 2.6 and tf nightly .Please, find the gist [**`here`**](https://colab.research.google.com/gist/kumariko/2a002d5a04f0877a56074474ab36d4df/51241.ipynb#scrollTo=RSMnc5XAg9Ba).Thanks!

**Comment by gowthamkpr** [comment]
@kumariko This has been fixed in the latest version of tensorflow (2.8.2 and 2.10 dev). Please find the gist [here](https://colab.research.google.com/gist/gowthamkpr/997316dfbe866f688ef5f765eff2468c/51241.ipynb). Thanks!
